### ISSUE TITLE
Perlmutter configuration not building

### ISSUE THREAD

**Issue by swelborn** [mochi-hpc-experiments/platform-configurations#51]
Hello. I'm wondering if someone could attempt to reproduce this bug on Perlmutter:

<details>
<summary>Spack install log</summary>

```

$ spack install        
==> Warning: cannot detect libc from gcc@=10.3.0. The compiler will not be used during concretization.
==> Warning: cannot detect libc from gcc@=11.2.0. The compiler will not be used during concretization.
==> Concretized mochi-margo
 -   77hv5ei  mochi-margo@0.18.3%gcc@12.3.0~pvar build_system=autotools arch=linux-sles15-zen3
 -   4h6mp5l      ^argobots@1.2%gcc@12.3.0~affinity~debug~lazy_stack_alloc+perf~stackunwind~tool~valgrind build_system=autotools stackguard=none arch=linux-sles15-zen3
[e]  dbmzbi4      ^autoconf@2.69%gcc@12.3.0 build_system=autotools patches=7793209 arch=linux-sles15-zen3
[e]  ixayhcy      ^automake@1.15.1%gcc@12.3.0 build_system=autotools arch=linux-sles15-zen3
 -   pgrxb4j      ^coreutils@9.4%gcc@12.3.0~gprefix build_system=autotools arch=linux-sles15-zen3
[+]  brkks4g      ^gcc-runtime@12.3.0%gcc@12.3.0 build_system=generic arch=linux-sles15-zen3
[e]  ea6wt6w      ^glibc@2.31%gcc@12.3.0 build_system=autotools arch=linux-sles15-zen3
[+]  bxti4iw      ^gmake@4.4.1%gcc@12.3.0~guile build_system=generic arch=linux-sles15-zen3
[+]  3u4nz3j      ^json-c@0.16%gcc@12.3.0~ipo build_system=cmake build_type=Release generator=make arch=linux-sles15-zen3
[e]  3irn5qe          ^cmake@3.30.2%gcc@12.3.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-sles15-zen3
[e]  wnw3gxd      ^libtool@2.4.6%gcc@12.3.0 build_system=autotools arch=linux-sles15-zen3
[e]  ebmezft      ^m4@1.4.18%gcc@12.3.0+sigsegv build_system=autotools patches=3877ab5,fc9b616 arch=linux-sles15-zen3
[+]  hfh4wbl      ^mercury@2.4.0rc5%gcc@12.3.0~bmi~boostsys~checksum~debug~hwloc~ipo~mpi+ofi~psm~psm2+shared+sm~ucx~udreg build_system=cmake build_type=Release generator=make arch=linux-sles15-zen3
[e]  3irn5qe          ^cmake@3.30.2%gcc@12.3.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=linux-sles15-zen3
[e]  om73h5v          ^libfabric@1.20.1%gcc@12.3.0~cuda~debug~disable-spinlocks~kdreg~uring build_system=autotools fabrics=rxm,sockets,tcp,udp arch=linux-sles15-zen3
[e]  x27s3js      ^pkg-config@0.29.2%gcc@12.3.0+internal_glib build_system=autotools arch=linux-sles15-zen3

[+] /usr (external glibc-2.31-ea6wt6wyyi3je7yf5vmcdg6chgi7bi6f)
[+] /usr (external autoconf-2.69-dbmzbi4z5b447zxtu5xypistbdy4pktk)
[+] /usr (external automake-1.15.1-ixayhcykfqaei6vcuqvszxziko5hufif)
[+] /usr (external libtool-2.4.6-wnw3gxdvmh6rri3chrlr434kjpgvtdwc)
[+] /usr (external m4-1.4.18-ebmezftkc5kqqlawu3rf2nbkia6j2lvj)
==> libfabric@1.20.1 : has external module in ['libfabric/1.20.1']
[+] /opt/cray/libfabric/1.20.1 (external libfabric-1.20.1-om73h5vontrcj2mz3trp2qwsnf5ojtwg)
[+] /usr (external pkg-config-0.29.2-x27s3jsneaskldeesktnyq3ttvihk7g4)
[+] /global/homes/s/swelborn/spack-workspace/perlmutter/software/linux-sles15-zen3/gcc-12.3.0/gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae
[+] /global/homes/s/swelborn/spack-workspace/perlmutter/software/linux-sles15-zen3/gcc-12.3.0/mercury-2.4.0rc5-hfh4wblu2tx5tugnsgvpcflfvst5n2th
[+] /global/homes/s/swelborn/spack-workspace/perlmutter/software/linux-sles15-zen3/gcc-12.3.0/gmake-4.4.1-bxti4iwawnm7ba76iin6qcnumhmfatud
==> Installing argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq [11/14]
==> No binary for argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq found: installing from source
==> Using cached archive: /global/homes/s/swelborn/.spack/cache/_source-cache/archive/1c/1c056429d9c0a27c041d4734f6318b801fc2ec671854e42c35251c4c7d0d43e1.tar.gz
==> No patches needed for argobots
==> argobots: Executing phase: 'autoreconf'
==> argobots: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq/spack-src/configure' '--prefix=/global/homes/s/swelborn/spack-workspace/perlmutter/software/linux-sles15-zen3/gcc-12.3.0/argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq' '--enable-perf-opt' '--disable-valgrind' '--disable-debug' '--enable-perf-opt' '--disable-valgrind' '--disable-debug' '--disable-lazy-stack-alloc'

2 errors found in build log:
     6     ==> [2024-12-10-07:22:38.606454] FILTER FILE: /global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq/spack-src/configure [replacing "^(\s*test x-R = )("\$p")(; t
           hen\s*)$"]
     7     ==> [2024-12-10-07:22:38.629745] FILTER FILE: /global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq/spack-src/configure [replacing "^(\s*test \$p = "-R")(; then
           \s*)$"]
     8     ==> [2024-12-10-07:22:38.656137] '/global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq/spack-src/configure' '--prefix=/global/homes/s/swelborn/spack-workspace/
           perlmutter/software/linux-sles15-zen3/gcc-12.3.0/argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq' '--enable-perf-opt' '--disable-valgrind' '--disable-debug' '--enable-perf-opt' '--disable-valgrind' '--disable-debug' '--disable-laz
           y-stack-alloc'
     9     checking whether the C compiler works... yes
     10    checking for C compiler default output file name... a.out
     11    checking for suffix of executables...
  >> 12    checking whether we are cross compiling... configure: error: in '/global/u1/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq/spack-src':
  >> 13    configure: error: cannot run C compiled programs.
     14    If you meant to cross compile, use '--host'.
     15    See 'config.log' for more details

See build log for details:
  /global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq/spack-build-out.txt

==> Warning: Skipping build of mochi-margo-0.18.3-77hv5ei6e2bgfhnclihxxnov4ynvim6k since argobots-1.2-4h6mp5lw3ya5nl2jtifjhgbacf5xuuvq failed
==> Installing coreutils-9.4-pgrxb4jhuon3bkmi7u2w47tixs3atiu3 [12/14]
==> No binary for coreutils-9.4-pgrxb4jhuon3bkmi7u2w47tixs3atiu3 found: installing from source
==> Using cached archive: /global/homes/s/swelborn/.spack/cache/_source-cache/archive/ea/ea613a4cf44612326e917201bbbcdfbd301de21ffc3b59b6e5c07e040b275e52.tar.xz
==> No patches needed for coreutils
==> coreutils: Executing phase: 'autoreconf'
==> coreutils: Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 77:
    '/global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-coreutils-9.4-pgrxb4jhuon3bkmi7u2w47tixs3atiu3/spack-src/configure' '--prefix=/global/homes/s/swelborn/spack-workspace/perlmutter/software/linux-sles15-zen3/gcc-12.3.0/coreutils-9.4-pgrxb4jhuon3bkmi7u2w47tixs3atiu3'

2 errors found in build log:
     11    checking whether make supports nested variables... yes
     12    checking whether make supports nested variables... (cached) yes
     13    checking for gcc... /global/common/software/spackecp/perlmutter/v0.22/96029/spack/lib/spack/env/gcc/gcc
     14    checking whether the C compiler works... yes
     15    checking for C compiler default output file name... a.out
     16    checking for suffix of executables...
  >> 17    checking whether we are cross compiling... configure: error: in '/global/u1/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-coreutils-9.4-pgrxb4jhuon3bkmi7u2w47tixs3atiu3/spack-src/spack-build':
  >> 18    configure: error: cannot run C compiled programs.
     19    If you meant to cross compile, use '--host'.
     20    See 'config.log' for more details

See build log for details:
  /global/homes/s/swelborn/spack-workspace/perlmutter/build_stage/spack-stage-coreutils-9.4-pgrxb4jhuon3bkmi7u2w47tixs3atiu3/spack-build-out.txt

[+] /global/homes/s/swelborn/spack-workspace/perlmutter/software/linux-sles15-zen3/gcc-12.3.0/json-c-0.16-3u4nz3jjvs42vvyvxwy6k5rvzgrvrg7w
==> Error: mochi-margo-0.18.3-77hv5ei6e2bgfhnclihxxnov4ynvim6k: Package was not installed
==> Updating view at /global/homes/s/swelborn/mochi/.spack-env/view
==> Error: Installation request failed.  Refer to reported errors for failing package(s).

```
</details>

Seems like we cannot run C programs?

Steps to reproduce: 

```sh
module purge # for clean module env
module restore # for default module env
module load spack
spack env create -d $HOME/mochi spack.yaml
spack env activate /global/homes/s/swelborn/mochi
spack repo add /global/homes/s/swelborn/gits/mochi-spack-packages
spack add mochi-margo
spack install
```

**Comment by mdorier** [comment]
The spack.yaml file is the one from this repo I assume?

**Comment by swelborn** [comment]
Yes under `NERSC/Perlmutter/`

**Comment by mdorier** [comment]
Ok for me it leads to this error:

```
[+] /usr (external glibc-2.31-ea6wt6wyyi3je7yf5vmcdg6chgi7bi6f)
[+] /usr (external autoconf-2.69-dbmzbi4z5b447zxtu5xypistbdy4pktk)
[+] /usr (external automake-1.15.1-ixayhcykfqaei6vcuqvszxziko5hufif)
==> cmake@3.30.2 : has external module in ['cmake/3.30.2']
[+] /global/common/software/nersc9/cmake/3.30.2 (external cmake-3.30.2-3irn5qeywas3kfohm5vexemv77v4v4bu)
[+] /usr (external libtool-2.4.6-wnw3gxdvmh6rri3chrlr434kjpgvtdwc)
[+] /usr (external m4-1.4.18-ebmezftkc5kqqlawu3rf2nbkia6j2lvj)
==> libfabric@1.20.1 : has external module in ['libfabric/1.20.1']
[+] /opt/cray/libfabric/1.20.1 (external libfabric-1.20.1-lu4dyoblnuyfbwgc3pv5xxbw2ptw5v7v)
[+] /usr (external pkg-config-0.29.2-x27s3jsneaskldeesktnyq3ttvihk7g4)
==> Installing gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae [9/14]
==> Fetching file:///global/common/software/spackecp/mirrors/perlmutter-v0.22/build_cache/linux-sles15-zen3-gcc-12.3.0-gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae.spec.json.sig
gpg: keybox '/global/homes/m/mdorier/.gnupg/pubring.kbx' created
gpg: Signature made Wed 22 May 2024 08:59:25 AM PDT
gpg:                using RSA key B5FDE18F615783AF078ED29C3BD6B0E9935AEB8F
gpg: Can't check signature: No public key
==> Warning: Failed to verify: file:///global/common/software/spackecp/mirrors/perlmutter-v0.22/build_cache/linux-sles15-zen3-gcc-12.3.0-gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae.spec.json.sig
==> Warning: Skipping build of argobots-1.2-jdgwg6dpevulcgs3acdhpdlt7mlmb2r6 since gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae failed
==> Warning: Skipping build of mochi-margo-0.15.0-37qtj6kj6jd5ktitl2yn5doexg7bkpuc since argobots-1.2-jdgwg6dpevulcgs3acdhpdlt7mlmb2r6 failed
==> Warning: Skipping build of json-c-0.16-3u4nz3jjvs42vvyvxwy6k5rvzgrvrg7w since gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae failed
==> Warning: Skipping build of gmake-4.4.1-bxti4iwawnm7ba76iin6qcnumhmfatud since gcc-runtime-12.3.0-brkks4glphmqjotqtuc652bbuh4t2qae failed
==> Warning: Skipping build of mercury-2.3.1-i7hwygzducedptufjnb2nqzf6u2o3pf3 since gmake-4.4.1-bxti4iwawnm7ba76iin6qcnumhmfatud failed
==> Error: Failed to install gcc-runtime due to NoVerifyException: Spack found new style signed binary packages, but was unable to verify any of them.  Please obtain and trust the correct public key.  If these are public spack binaries, please see the spack docs for locations where keys can be found.
==> Error: mochi-margo-0.15.0-37qtj6kj6jd5ktitl2yn5doexg7bkpuc: Package was not installed
==> Updating view at /global/homes/m/mdorier/test-env/.spack-env/view
==> Error: Installation request failed.  Refer to reported errors for failing package(s).
```

I haven't used Perlmutter in years. @carns do we have anything recent running there? Have you seen this kind of error?

**Comment by mdorier** [comment]
I tried with a fresh clone of spack (as opposed to Perlmutter's module) and it built some of the dependencies (e.g. json-c and mercury) but failed with argobots with a similar error.

**Comment by carns** [comment]
I've never tried using a facility-provided spack instance, but as a sanity check I started with my own clone of spack.  It works fine for me (see attached log).

[perlmutter-spack-example-2024-12-10.log](https://github.com/user-attachments/files/18087794/perlmutter-spack-example-2024-12-10.log)

I don't know off hand what's different about the spack provided via modules on Perlmutter, but I'll give it a try and see what happens.

**Comment by carns** [comment]
I just tried using the NERSC-provided Spack module.  It didn't work for me either, but it matched @mdorier 's error rather than @swelborn 's error (see attached log).

[perlmutter-system-spack-debugging-2024-12-10.log](https://github.com/user-attachments/files/18087894/perlmutter-system-spack-debugging-2024-12-10.log)

gcc-runtime is not provided by us. This is failing to verify something from the local buildcache at NERSC, I think.  Is there some other configuration step needed?

**Comment by carns** [comment]
Ah, Ok, I was able to progress to @swelborn 's error by using `spack install --use-buildcache never`.   Once I did that I got this failure on the Argobots build:

```
  >> 12    checking whether we are cross compiling... configure: error: in '/gl
           obal/u1/p/pcarns/spack-workspace/perlmutter/build_stage/spack-stage-
           argobots-1.2-r45vwsebsqfdkosq6nd7mvuexz5c6ofw/spack-src':
  >> 13    configure: error: cannot run C compiled programs.
     14    If you meant to cross compile, use '--host'.
     15    See 'config.log' for more details
```

Looking in the config.log in ` /global/u1/p/pcarns/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-r45vwsebsqfdkosq6nd7mvuexz5c6ofw/spack-src` it shows this as the underlying problem:

```
configure:3332: ./conftest
./conftest: error while loading shared libraries: libcudart.so.12: cannot open shared object file: No such file or directory
configure:3336: $? = 127
configure:3343: error: in '/global/u1/p/pcarns/spack-workspace/perlmutter/build_stage/spack-stage-argobots-1.2-r45vwsebsqfdkosq6nd7mvuexz5c6ofw/spack-src':
configure:3345: error: cannot run C compiled programs.
If you meant to cross compile, use '--host'.
See 'config.log' for more details
```

The `libcudart.so.12: cannot open shared object file: No such file or directory` is the crucial bit.

I don't know for sure, but I think this is provided by the `cudatoolkit` module on Perlmutter.  In our spack environment we specify that package as a required module for the compiler, see https://github.com/mochi-hpc-experiments/platform-configurations/blob/d90bf78a88d414388d03e44cb9a1f971365bbcef/NERSC/Perlmutter/spack.yaml#L25 .

Looking at the output of `spack config blame` it looks like the NERSC spack has provided its own compiler configuration.  This one might be overriding ours (I'm not sure how to tell exactly which one it is using) and it does not have cudatoolkit as a module:

```
                               spec: gcc@12.3.0
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:81                                   paths:
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:82                                     cc: /opt/cray/pe/craype/default/bin/cc
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:83                                     cxx: /opt/cray/pe/craype/default/bin/CC
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:84                                     f77: /opt/cray/pe/craype/default/bin/ftn
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:85                                     fc: /opt/cray/pe/craype/default/bin/ftn
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:86                                   flags: {}
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:87                                   operating_system: sles15
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:88                                   target: any
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:89                                   modules:
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:90                                   - PrgEnv-gnu
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:91                                   - gcc-native/12.3
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:92                                   - craype-x86-milan
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:93                                   - libfabric
/global/common/software/spackecp/perlmutter/spack_settings/compilers.yaml:94                                   extra_rpaths: []
```

Maybe we need to force it to use our compiler spec, or else get the NERSC one updated to load the CUDA stuff?  I'm not sure how to proceed from here, but I think the underlying problem.

**Comment by mdorier** [comment]
Ok thanks to @carns I think I solved it: yesterday when I tried, I was getting @swelborn's error. Then I tried with my own installation of Spack and I was still getting the error. Today I tried again my own installation of Spack, and it works fine. The difference is that today I didn't do `module load spack`. Turns out even if you use your own installation of Spack, if you also do `module load spack`, you end up with your Spack picking up both your own configuration and the system's spack configurations.

TL;DR: don't load Perlmutter's spack module, clone your own Spack and do `source spack/share/spack/setup-env.sh`.
